### ISSUE TITLE
Aggregator performance ?

### ISSUE THREAD

**Issue by RobinDevTW** [grobian/carbon-c-relay#198]
Hello,

I'm migrating our carbon-python infrastructure to carbon-c. 
As we were previously using the pickled port, and we cannot redeploy all our processes at once, we have a transition period with both carbon-python and python-c. 

Currently, we set up a carbon-c-relay taking most of the metrics. Then it is redirecting to carbon-python aggregator or another carbon-python relay which will write directly to carbon cache.
It works well.  So now, we wanted to give a try to the carbon-c aggregator, by switching only a small part of our metrics. 
So we wrote the aggregate regex to compute different sum/avg and added a match rule in order to not send those metrics to the carbon-python aggregator. (do you still follow me ?)

So it seems that the metrics are well redirected to the aggregator, but half of the aggregated values are lost. Look at this graph: http://i.imgur.com/RPdWhh0.png
Before 15:00, it's the carbon-c aggregator [every 60s - expire after 180s]
Between 15:00 and 15:45: we rolled back to carbon-python aggregator
Between 15:45 and 16:15: carbon-c aggregator [every 60s - expire after 80s]
After 16:15: rollback to carbon-python

It looks like there is something blocking or slowling down the metrics. Funny thing, when stopping or reloading the config, some "old" (like 10-20min old) aggregated metrics are coming and filling up some part of the graphs.
Any idea what's happening and how to solve this issue ?

For information , we are talking about 1M metrics/ min and we selected only 200k to be aggregated for our test. But at the end, we will aggregate all the metrics (~20 metrics = 1 aggregated metric).

Thanks

**Comment by grobian** [comment]
can you shared your (annonymised) aggregation rules?  I suspect you run into a known problem that still needs a solution where many distinct aggrgations are involved.

**Comment by RobinDevTW** [comment]
match ([^.]+)\.nodes\.([^.]+)\.[^.]+\.abc\.(.+)
  send to
    relay_b
;

aggregate ([^.]+)\.nodes\.([^.]+)\.[^.]+\.abc\.(.+)
  every 60 seconds
  expire after 180 seconds
  compute sum write to \1.agg_sum.nodes.all.abc.\3
  compute average write to \1.agg_avg.nodes.all.abc.\3
  compute sum write to \1.agg_sum.nodes.cluster.\2.abc.\3
  compute average write to \1.agg_avg.nodes.\2.abc.\3
  send to graphite
  stop
;

**Comment by grobian** [comment]
right, and how many unique metrics do you expect as output?  (the computes)

**Comment by RobinDevTW** [comment]
For 200k metrics, we expect about 10k  unique aggregated metrics.
This is for our current test. But when we will aggregate everything, it will be 1M metrics => about 50k unique aggregated metrics.

**Comment by RobinDevTW** [comment]
I compared the metrics produced by :
(1) compute sum write to \1.agg_sum.nodes.all.abc.\3
and
(2) compute sum write to \1.agg_sum.nodes.\2.abc.\3

For a same "metric", the metrics aggregated in (2) look (at first sight) ok. I didn't check all of them though. While the aggregated values from rule (1) are scattered (and (1) is simply a sum of all (2) for a same rule)

**Comment by grobian** [comment]
hmmm, I need to dive into there, it may have to do with the fact that (1) has more "writes" than (2), but it seems obvious this needs to be done completely different.

**Comment by RobinDevTW** [comment]
Then, maybe (1) should be rewritten with a new aggregate statement that will aggregate the aggregated data from (2) ?
Like:
`aggregate ^([^.]+)\.agg_sum.nodes\.[^.]+.abc\.(.+)
	compute sum write to \1.agg_sum.nodes.all.abc.\2`

Or... the order of the "compute" statement is important ?

I actually did it in one single aggregate rule because the document was stating that additional compute rules "come for free".

**Comment by grobian** [comment]
right, they "come for free" given the single regexp match, however, when I implemented the first version of the aggregator noone was throwing million of metrics per minute at it, there's some problems that seem to boil up now that need addressing.

**Comment by RobinDevTW** [comment]
I have done some other experiments with some findings. 
Currently, the aggregated values were the only which were sent directly to the carbon caches. All other metrics were sent to another python-relay before going to carbon caches.
I wanted to be sure that running another python-aggregator was not messing up the values.
So I gave a new temp name to the output aggregated values.
Observations:
- I took a very long time (>10 minutes) before the new metrics were created
- Some metrics were not even created after 20 minutes
- The aggregated values were off by 60%

Then I reloaded a new configuration.
Observations:
- All the aggregated values are suddenly created (well suddenly => ~5minutes)
- The 10 past minutes aggregated values that were completely off are progressively corrected (took around 10 minutes)
- Older aggregated values are not corrected

So there must be some kind of queue/cache which is not flushed at runtime but when stopping the process (or reloading a configuration), it is flushed  ?

Next experiment is then to use python-relay as destination of the aggregated values, instead of directly carbon-caches.
Now, the aggregated values are similar to the manual sum of the values !!
 
So maybe the issue is not related to aggregation but to the cluster of carbon caches ?
Here is the configuration:

```
cluster graphite
  carbon_ch
    127.0.0.1:2103=a
    127.0.0.1:2203=b
    127.0.0.1:2303=c
    127.0.0.1:2403=d
    127.0.0.1:2503=e
    127.0.0.1:2603=f
;

```

**Comment by RobinDevTW** [comment]
Long story, short.
We found the issue. Among the thousands of server we updated with the new port to C-Relay, one of them was still using the Python-Relay and therefore the python-aggregator.
This one was overriding the metrics sent by C-Relay. 

So I don't think we have aggregator performance issue, at the moment at least.
Still doesn't explain the issue when sending metrics directly to the caches. We'll try that again later.
However, it does not relate to this ticket anymore, we'll open a new ticket if needed.

We can close this ticket. Thanks for your support and sorry for the convenience.
